getinsomnia/insomnia
# UI for filter requests by folder, URI, method, query string

Related PRs?
-------------
#797 - I previously implemented some of the logic to allow searching by folder, URI, and query string here, but nothing was apparent in the UI.
#857 - Continuation of #797, for implementing the UI of the fuzzy matching results and for adding URI to the Quick Switch window rows.

Relevant Issues?
-----------------
#739 - This extends #797 in fulfilling #739 by providing a Highlight component that has been added to the Quick Switch window.

What does this PR do?
-----------------------
* Replaces the fuzzy matching functions with the [fuzzysort](https://github.com/farzher/fuzzysort) library, as it provides similar functionality but also provides highlighting capabilities.
* Creates a reusable Highlight component that shows the matched characters in the fields searched against.
* Adds that Highlight component to the Quick Switch window fields, and adds URL with querystring to Quick Switch results as well.

Example screenshots, gifs, or videos?
--------------------------------------
* Screenshot: https://cl.ly/0p1i2c2e3N18
* GIF: https://cl.ly/011H001a2J3k

What is left to do or discuss?
-----------------------------
* What should we do with filtering on the sidebar? I didn't implement any UI for it yet, and I haven't added Method or URI/querystring back in for it. I'm thinking we leave method/URI/querystring out as it wouldn't likely fit. But I might at least add the Highlight component to the sidebar to make it more apparent what filters matched.
* I have an inline style or two, looks like most of the codebase doesn't use that, so I can extract to a class.
* The Highlight component works fairly well. The only thing I see is that fuzzysort doesn't seem to match multiple times for one of the phrases. I'm assuming it is fuzzysort and not my code...maybe fuzzysort has/needs a /g type of option? https://cl.ly/121S3F3U2z11
* I need to truncate or wrap long URIs in the Quick Switch window. Long URIs go to the edge of the modal at the moment.

What code can others reuse?
-----------------------------
* Highlight component - We can add the fuzzy matching UI pretty easily anywhere else in the UI we want to implement something similar. Just give it a fuzzy matching string and the text that you want to match/display.

Testing Steps
--------------
* Search for URI, method, query string, folder, and/or request name on Quick Switch window
* Perform fuzzy search, leaving out some letters for a given field
* Perform a multiple word search, searching multiple types of fields in the list above
* Repeat tests for Sidebar search - nothing should have changed, except the fuzzy search algorithm is now using the [fuzzysort](https://github.com/farzher/fuzzysort) library
